apps/rush-lib/src/logic/setup/KeyboardLoop.ts

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as readline from 'readline';
+import * as process from 'process';
+
+export class KeyboardLoop {
+  protected stdin: NodeJS.ReadStream;
+  protected stderr: NodeJS.WriteStream;
+  private _readlineInterface: readline.Interface | undefined;
+  private _resolvePromise: (() => void) | undefined;
+  private _rejectPromise: ((error: Error) => void) | undefined;
+  private _cursorHidden: boolean = false;
+
+  public constructor() {
+    this.stdin = process.stdin;
+    this.stderr = process.stderr;
+  }
+
+  public get capturedInput(): boolean {
+    return this._readlineInterface !== undefined;
+  }
+
+  private _captureInput(): void {
+    if (this._readlineInterface) {
+      return;
+    }
+
+    this._readlineInterface = readline.createInterface({ input: this.stdin });
+
+    readline.emitKeypressEvents(process.stdin);
+    this.stdin.setRawMode!(true);
+    this.stdin.addListener('keypress', this._onKeypress);
+  }
+
+  private _uncaptureInput(): void {
+    if (!this._readlineInterface) {
+      return;
+    }
+
+    this.stdin.removeListener('keypress', this._onKeypress);
+    this.stdin.setRawMode!(false);
+    this._readlineInterface.close();
+    this._readlineInterface = undefined;
+  }
+
+  protected hideCursor(): void {
+    if (this._cursorHidden) {
+      return;
+    }
+    this._cursorHidden = true;
+    this.stderr.write('\u001B[?25l');
```

Comment:
Can you stick these control characters in constants with meaningful names?

Reply:
Done. I want to integrate this into the node-core-library `AnsiEscape` APIs, but when I started thinking about that, it goes beyond the scope of this PR.